Refactor/nomenclature correction

antoniojefferson opened the pull request:
Renomeação de esquemas de models e controllers;
Renomeção do projeto;
Remoção do controllere addresses quenão tem uso;

antoniojefferson committed:
refactor(application.rb): renomeando projeto

antoniojefferson committed:
refactor(addresses): renomeacao da tabela adresses para addresses

Foi renomeado todo o esquema do endereço que antes estava domo adress, agora ficou address, foi
renomeado assim todo o esquema de model, controller e tests

antoniojefferson committed:
refactor(citizens): renomeando tabela municipes para citizens

Foi renomeado todo o esquema do cidadao comum que antes estava como municipe, agora ficou citizen,
foi renomeado assim todo o esquema de model, controller

antoniojefferson committed:
fix(addresses): renomenado a foreign key de citizen

Como foi atualizando o esquema d cidadao comum, onde o mesmo estava como municipe e agora e citizen,
a foreign key nao havia sido renomeada para a nomenclatura correnta, onde ela deixa de ser
municipe_id para ser citizen_id

antoniojefferson committed:
refactor(routes.rb): corrigindo as rotas para o novo esquema

Corrigindo as rotas para a nova nomenclatura no esquema de models do projeto.

antoniojefferson committed:
refactor(schema.rb): refatoraçao ocorrida devido a mudança de nomes das tables

antoniojefferson committed:
refactor(routes.rb): removendo rotas desnecessarias de addresses

Removendo as rotas desnescessarias de addresses, por nao haver nescessidades para ela uma vez
atualmente toda a informacao e retornada via o controller de citizens

antoniojefferson committed:
refactor: removendo arquivoes desnecessarios

ricardopacheco commented on app/controllers/citizens_controller.rb:
```diff
@@ -0,0 +1,54 @@
+class CitizensController < ApplicationController
+  before_action :find_citizen, except: %i[create index]
+
+	# GET /citizens
```
Há discrepância na configuração do seu editor e do git quando a questão de tab/space. Peço que veja isso, para que o código escrito no arquivo de fato fique igual conforme o git registra.

Usamos identação (2) com spaces por padrão.

marcelohenriq99 replied:
Vai nas configurações e seta o valor `2` no tab size e tira o auto detect pra ele sempre acusar se tiver mais espaços

antoniojefferson replied:
Meu vs code já esta configurado para 2 espaços, ele subu o código assim por que, antes não estava, antes estava em 4 espaços, quando setei para 2 espaços ele reorganizei o código ele parece ter deixado uma especie de tab ali, mas que no vs code fica a largura de 2 espaços.

antoniojefferson replied:
Vou corrigir aqui e subir uma correção

ricardopacheco commented on app/controllers/citizens_controller.rb:
```diff
@@ -0,0 +1,54 @@
+		@citizens = Citizen.all
+		@citizens.each do |citizen|
+			@finalResult.push(citizen.json)
+		end
```
1. Dá pra fazer esse código mais legível e otimizado com: `Citizen.all.map(&:to_json)`
2. Nem o código acima é necessário:

```
def index
  citizens = Citizen.all
  
  render json: citizens, status: :ok
end
```

Porque funciona? O ActionController entende pelo renderer que você quer renderizar um content-type `application/json`, e, já faz isso pra ti maroto. Formatos como html/json (xml n me recordo) ele faz por padrão. Em caso de outros formatos, seria necessário configurar o mimetype da app (csv,xls) para o rendereder entender (ou, escrever manualmente na action).

antoniojefferson replied:
Show, vou estar atulizando esse código!

ricardopacheco commented on app/controllers/citizens_controller.rb:
```diff
@@ -0,0 +1,54 @@
+		@citizen = Citizen.new(citizen_params)
+		if @citizen.save
+
+			render json: @citizen, status: :created
```
Olha ele fazendo aqui maroto o que comentei acima! ;)

ricardopacheco commented on app/controllers/citizens_controller.rb:
```diff
@@ -0,0 +1,54 @@
+
+	def find_citizen
+		@citizen = Citizen.find(params[:_id])
+		rescue ActiveRecord::RecordNotFound
```
Essa é uma abordagem teoricamente correta, mas, devido a questões de performance o uso de exception handling não é considerado boa prática (exceto no caso de large apps que precisem mapear erros). Em apps pequenas é desperdicio, mas, em apps grandes é uma excelente ideia (o rails cria classes de erro para especificar exceptions. Exemplo: https://www.rubydoc.info/docs/rails/4.1.7/ActiveRecord/ActiveRecordError

Minha recomendação seria:

```
def find_citizen
  @citizen = Citizen.find_by(id: params[:id])
  
  render json: I18n.t('errors.messages.citizen_not_found') if @citizen.blank?
end
```

PS: `find_by` retorna `nil` quando o record não é encontrado. Vc pode usar o mesmo comportamento de find usando bang `(find_by!)`

antoniojefferson replied:
Top demais, gostei da sua sugestão, considere feita mano!

marcelohenriq99 replied:
esse método ai deveria mesmo ter um `render` ? ele não deveria somente retornar o `municipe` ou `nil` caso não encontre ?
o tratamento disso ficaria nas `actions` do `controller`

italopires replied:
removeria o render, que pra esse caso não é necessário já que é chamado em um `before_action`

E como vai ser buscado apenas por `id`, usaria so find mesmo
```
@citizen = Citizen.find(params[:id])
```

marcelohenriq99 replied:
Cuidado se for usar o `find` ele levanta uma exception se não achar o registro, ja o `find_by` retorna nil, eu não vejo problema em usar o `find_by`

antoniojefferson replied:
O render fica ai pelo fato de esse metodo, ser usado para buscar um cidadão por id no contexto dos metodos show e update, o render nessa função deixa uma unica linha de retrno de erro caso não encontre o id procurado, removendo ele dai eu precisaria retornar um erro para o caso onde ele não encontre o id, o que faria com que eu adicionasse um linha para retornar o erro no show e no update, ficaria repetitivo.

ricardopacheco commented on app/models/address.rb:
```diff
@@ -1,5 +1,5 @@
-class Adress < ApplicationRecord
-  belongs_to :municipe, optional: true
+class Address < ApplicationRecord
```
👍

marcelohenriq99 commented on app/models/citizen.rb:
```diff
@@ -14,7 +14,7 @@ def json
         photo: self.photo,
         status: self.status,
-        adress: self.adresses,
+        address: self.addresses,
```
Se tua relation é um `has_many` não seria mais correto `addresses: self.addresses`

antoniojefferson replied:
Sim, foi vacilo meu, esqueci de pluralizar

ricardopacheco replied:
Bem lembrado marcelo!

ricardopacheco commented on app/models/address.rb:
```diff
@@ -1,5 +1,5 @@
-class Adress < ApplicationRecord
-  belongs_to :municipe, optional: true
+class Address < ApplicationRecord
+  belongs_to :citizen, optional: true
```
💥  Atenção quanto a regra de negócio respondendo a seguinte pergunta: 

Faz sentido ter endereços cadastrados no banco sem um cidadão associado?

marcelohenriq99 replied:
no Rails >= 5 ficou obrigatório os relacionamento `belongs_to` por isso esse`optional: true`, sem isso a validação falha por esperar que já exista um endereço cadastrado.

ricardopacheco replied:
@marcelo-henriq o que você disse é um fato verdadeiro, porém, não é esse o contexto.

Vou aguardar nosso novo dev para então eu responder ;)

antoniojefferson replied:
Então, concordo que não faz sentido, isso já vai ser corrigido, esse ponto passou despercebido, eu estava fazendo um teste junto com o Marcelo, durante o teste estavam havendo alguns problemas de cadastro, daí eu adicioneui esse opção para conseguir relizar o fluxo, pela logica essa opção não faz sentido, estarei corrigindo e subindo a atualização com esta opção removida.

marcelohenriq99 commented on config/routes.rb:
```diff
@@ -1,14 +1,9 @@
-  post '/municipes', to: 'municipes#create'
-  put '/municipes/:_id', to: 'municipes#update'
+  # Routes citizens
+  get '/citizens', to: 'citizens#index'
```
cria essas rota com o `resources`, que é o padrão do rails

antoniojefferson replied:
Certo, vou subistuir

marcelohenriq99 commented on test/models/address_test.rb:
```diff
@@ -1,6 +1,6 @@
 require "test_helper"
 
-class AdressTest < ActiveSupport::TestCase
+class AddressTest < ActiveSupport::TestCase
```
Aqui nos testes, procura usar o [rspec](https://github.com/rspec/rspec-rails)

antoniojefferson replied:
Esse é opadrão projeto, apenas renomeei, ainda não inclui os códigos

marcelohenriq99 replied:
isso é por conta dos generators do rails, tem que ver isso tbm, os generators criam alguns arquivos desnecessários, tem como skipar isso com algumas flags da uma olhada depois, mas eu aconselharia a não se a pegar aos generators, cria os arquivos na mão msm, assim não acontece essas coisinhas.

ricardopacheco replied:
Excelente resposta marcelo, é exatamente isso!

ricardopacheco commented on app/models/citizen.rb:
```diff
@@ -14,7 +14,7 @@ def json
         photo: self.photo,
         status: self.status,
-        adress: self.adresses,
+        address: self.addresses,
```
Boa! Só lembrando que, conforme o time lembrou a vc no PR anterior, fazemos a serialização dos dados em classes específicas (serializers) ao invés de utilizarmos o ActiveModel/ActiveRecord diretamente no model. No caso EXTREMO de ser necessário tratar a nível de model (é bad practice, exceto um motivo dos bão), você poderia usar os métodos do ActiveRecord [as_json](https://api.rubyonrails.org/classes/ActiveModel/Serializers/JSON.html) e [to_json](https://apidock.com/rails/ActiveRecord/Serialization/to_json).

Porque é bad practice? Porque quebra o princípio de responsabilidade única. Models devem ser tratamento de dados e constraints orientados ao banco de dados, não a formatação de dados.

antoniojefferson replied:
Sim, esse ponto é o que vou tratar agora, não criei os serializers nessa PR por que queria deixar as duas bem separadas essa sendo para renomeação de tabelas e arquivos e a proxima que é para ser esse tratamento.

O intuito foi deixar esses dois processos separados, tentano manter uma organização, não consigo afirmar com certeza se minha linha de pensamento esta correta, mas foi uma ideia para deixar as PR organizadas e poder praticalas ao mesmo tempo.

marcelohenriq99 commented on test/models/citizen_test.rb:
```diff
@@ -1,6 +1,6 @@
-class MunicipeTest < ActiveSupport::TestCase
+class CitizenTest < ActiveSupport::TestCase
   # test "the truth" do
   #   assert true
```
Cuidado com os comentários no código

antoniojefferson replied:
Isso foi criado pelo proprio projeto, deixei por ter sido criado de padrão enquanto não modifiquei o arquivo para incluir os códigos de teste, mas assim que incluir os códigos esses comentarios não mais existirão

ricardopacheco commented on config/routes.rb:
```diff
@@ -1,14 +1,9 @@
-  post '/municipes', to: 'municipes#create'
-  put '/municipes/:_id', to: 'municipes#update'
+  # Routes citizens
+  get '/citizens', to: 'citizens#index'
```
Porque seguir assim?

`resources :citizens, except: :destroy` criaria pra você todas as rotas, exceto a de deletar. Consulte o guides do rails na seção `Routing` que isso é dado em detalhes.

antoniojefferson replied:
Já corrijo esse ponto

ricardopacheco commented on db/migrate/20211215172655_rename_municipes_table_name_to_citizens.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameMunicipesTableNameToCitizens < ActiveRecord::Migration[6.1]
```
Show!

marcelohenriq99 commented on app/models/citizen.rb:
```diff
@@ -1,7 +1,7 @@
-    accepts_nested_attributes_for :adresses, allow_destroy: true
+    accepts_nested_attributes_for :addresses, allow_destroy: true
 
     def json
```
Tenta encaixar o `serializer` pra tirar esse método daqui

antoniojefferson replied:
Serializer é na proxima PR, essa foi feita com o intuito apenas de renomear todo o processo e deixar tudo ok, para assim ser trabalhado em uma PR separada todo o tratamento de dados e seus retornos.

marcelohenriq99 commented on app/controllers/citizens_controller.rb:
```diff
@@ -0,0 +1,54 @@
+	def find_citizen
+		@citizen = Citizen.find(params[:_id])
+		rescue ActiveRecord::RecordNotFound
+			render json: { errors: 'Citizen not found' }, status: :not_found
```
Coloca `I18n` nessas mensagens de erro, caso precise exibir em tela fica mais fácil e é o padrão

antoniojefferson replied:
Pode deixar

ricardopacheco replied:
Remover o render nesse caso depende do propósito da aplicação ou do time de desenvolvimento/produto.

1. Se é um caso onde o produto entende que DEVE DAR ERRO SE NAO OCORRER COMO DEVERIA, o certo é deixar como `find`, e,  a exception será levantada, nesse caso acontecendo um 500 (internal server error). Em rails apps monolito isso acaba sendo tratado e um crud default pode retornar 404 (mas a exception será levantada de qualquer forma).

2. No caso de resource not found (404), quase sempre queremos tratar o erro. Usar rescue é correto também, porém, o código que mostrei é mais otimizado em termos de memória além de flexibilizar o http status.

3. Há uma diferença entre retornar 500 e 404, tanto a back quanto front. No entendimento geral, 404 só não foi encontrado, enquanto 500 quer dizer que algo que não deveria acontecer, aconteceu. Um exemplo disso é no frontend SPA onde o comportamento dos componentes podem mudar de acordo com o http status (`.then((response) =>  setState(response.data)).catch(error => throw Error(error))` sendo um caso rápido de onde o http status IMPORTA.

Resumindo, fazendo um if com find_by e retornando o json com http status not_found é simples e eficiente, na minha opinião.

marcelohenriq99 replied:
@ricardopacheco show demais, mas o que a gente quis dizer é que o metodo `find_citizen` como diz o nome só deveria se preocupar com encontrar ou não o municipe e retornar, o tratamento disso em conjunto com o que renderizar ficam mais entendíveis e mais corretos dentro das actions do controller, que é onde esse tipo de fluxo é justamente controlado, por isso a sugestão de remover o render desse metodo.

italopires replied:
Vejam a linha 2, temos um before_action, se perguntem, por que adicionar um `render` dentro de um metodo que usa `before_action`?

antoniojefferson replied:
É o seguinte o metodo render está ai pelo fato de esse metodo, ser usado para setar um cidadão por id no contexto dos metodos show e update, caso ele encontre show de bola, ele passa e continua o processo, caso ele não encontre, quero poder retornar ao front uma mensagem orientando o que aconteceu, se ele achou ele apresenta o show ou completa o update normal, caso ele não encontre nos dois casos eu preciso apresentar uma mensagem de retorno para orientar que o motivo não fo um erro e sim a não identificação de tal id, como a mensagem seria a mesma, faz sentido ela ser mostrada or esta função, ou seja bateu nela achou segue, não achou idenpendente se show ou update ele já renderiza a mensagem, removendo ele dessa função eu teria que tratar o mesmo erro duas vezes uma para metodo show e uma para o metodo update, ao meu ver seria repetivo e desnecessario, pois esta forma soluciona o problema resove tudo com duas linhas só.

antoniojefferson committed:
refactor(citizens_controller): corrigindo os espacos para 2, onde hoje esta como sendo uma tab

italopires commented on db/migrate/20211215172655_rename_municipes_table_name_to_citizens.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameMunicipesTableNameToCitizens < ActiveRecord::Migration[6.1]
+  def change
```
Nesse caso para renomear nome de tabela, precisa utilizar com os metodos `up` e `down`.

Porque o rails não tem como saber qual o nome da tabela era antes.

O método `change` deve ser utilizado apenas em migrations simples, como criacao de novas colunas, e o rollback é facilmente decifrado pelo rails, que é remover a coluna apenas.

Obs: faria essas mesmas mudancas nas migrations abaixo também

ricardopacheco replied:
> Porque o rails não tem como saber qual o nome da tabela era antes.

Essa afirmação é imprecisa. O que podemos dizer de fato é que o ActiveRecord para a maioria majoritária dos casos sabe reverter as migrations automaticamente. Usar `up` e `down` é em casos de migrações complexas que o mesmo não sabe como lidar (casos que precisem usar reversible no braço é um exemplo de quando o AR não sabe lidar e precisamos ir pro up/down).

antoniojefferson replied:
Show demais sua explicação @ricardopacheco

italopires replied:
👀

marcelohenriq99 reviewed (commented):
Tem que ver tbm o titulo e a descrição do PR, pra não misturar dois idiomas.

antoniojefferson committed:
refactor(citizen.rb): corrigindo a pluralizacao do atributo address

antoniojefferson committed:
refactor(routes.rb): simplificando as rotas, usando o resources

antoniojefferson committed:
refactor(test/*): removendo comentarios dos arquivos de test

antoniojefferson committed:
refactor(citizens_controller.rb): otimizando o codigo no metodo index no controller

antoniojefferson committed:
feat(pt-br.yml): adicionando arquivo de traduçoes, para uso no i18n

antoniojefferson committed:
feat(locale.rb): adicionando arquivo de config, para incializacao do i18n

antoniojefferson committed:
refactor(citizen_controller.rb): refatorando retorno de erro para falta de dados

Refatorando o erro, para caso ele nao encontre dados no metodo find_citizen, essa refatoracao inclui
o recurso de i18n.

antoniojefferson committed:
refactor(citizens_controller.rb): refatorando codigo seguindo sugestoes da PR

Aplicando i18n; Simplificando codigos; Melhorando apresentaçao de erros;

antoniojefferson committed:
refactor(citizen.rb): apicando validates para os campos do model

antoniojefferson committed:
feat(locale.rb): criando arquivo de inicializacao do i18n

antoniojefferson committed:
feat(pt-br.yml): criando configuracoes globais do i18n

antoniojefferson committed:
feat(citizen.yml): criando arquivo para config do i18n para model citizen

antoniojefferson committed:
refactor(citizens_controllers.rb): corrigindo comentario de identificacao da rota no verbo PUT

antoniojefferson committed:
refactor(address.rb): adicionando validate para os campos obrigatorios

antoniojefferson committed:
feat(address.yml): criando configs do i18n para o model address

antoniojefferson committed:
refactor(readme.md): corrigindo titulo e informaçoes do arquivo

antoniojefferson committed:
refactor(address.rb): removendo campo citizen_id do validate

Foi removido o campo citizen_id, inicialmente adicionado no validate do model address por estar
inserferindo no insert de novos registros.

italopires commented on app/models/citizen.rb:
```diff
@@ -1,7 +1,8 @@
-class Municipe < ApplicationRecord
-    has_many :adresses
+class Citizen < ApplicationRecord
+    has_many :addresses
```
Verifica a identação desse arquivo.

Outra coisa (opciohnal): pode ver isso mais pra frente, mas como na tua aplicação, pra cada municipe tem apenas um endereço, isso aqui ficaria apenas `has_one :address`.

antoniojefferson replied:
Já corrigi a identação

antoniojefferson replied:
Vou deixar o has_one então

antoniojefferson committed:
refactor(address.rb): romovendo opcao optional do belongs_to

Esta opcao nao e nescessaria, esta sendo removida, por que apesar de estar funcinando com ela, nao e
a forma correta de se utilizar e foge a regra de negocio

italopires commented on db/migrate/20211215172740_rename_adresses_table_name_to_addresses.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameAdressesTableNameToAddresses < ActiveRecord::Migration[6.1]
+  def change
```
👀

italopires commented on db/migrate/20211215185507_rename_municipe_field_to_addresses.rb:
```diff
@@ -0,0 +1,5 @@
+class RenameMunicipeFieldToAddresses < ActiveRecord::Migration[6.1]
+  def change
```
👀

antoniojefferson committed:
refactor: corrigindo identacoes

antoniojefferson committed:
Merge branch 'refactor/nomenclature_correction' of https://github.com/antoniojefferson/registration into refactor/nomenclature_correction

Megeando codigos

antoniojefferson committed:
refactor: toranando a relacao de address e citizen, uma para uma

antoniojefferson committed:
refactor(migrations): adicionando os metodos up e down nas migrations

Mudando as migracoes de atualizacao de nomes das tables e na mudanca da foreign key para usar os
metodos de up e down.

italopires reviewed (approved):
Perfeito 🚀

marcelohenriq99 reviewed (approved):
:rocket:
